### ISSUE TITLE
Feature request: add parameter to list all *.dex files from *.oat/odex without disassembling it.

### ISSUE THREAD

**Issue by bursoft-git** [JesusFreke/smali#376]
JesusFreke, I have a question:
How can I find out how many classes*.dex are in a *.odex without disassembling it?

For example, "java -jar baksmali-2.1.1.jar GmsCore.odex" output is:
```
GmsCore.odex contains multiple dex files. You must specify which one to disassemble with the -e option
Valid entries include:
/system/priv-app/GmsCore/GmsCore.apk
/system/priv-app/GmsCore/GmsCore.apk:classes2.dex
/system/priv-app/GmsCore/GmsCore.apk:classes3.dex
```

But, for "java -jar baksmali-2.1.1.jar TimeService.odex" output is:
`Warning: You are disassembling an odex file without deodexing it. You won't be able to re-assemble the results unless you deodex it with the -x option`
AND "out" folder with disassembled *.smali files.

In my batch deodex tool it wery slows deodexing process.
It is possible to add some sort of "-list" parameter to baksmali?

Thanks.

**Comment by JesusFreke** [comment]
Yeah, I've been meaning to put something like this in. I kinda want to redo the whole command line interface, tbh. It's getting a bit old/crufty :)

**Comment by bursoft-git** [comment]
It would be great!

**Comment by bursoft-git** [comment]
JF, I'm sorry for importunity
Is there any progress in this?

**Comment by JesusFreke** [comment]
Actually yep :) The next released version will have a new command line interface, and will include commands to list the dex files

e.g.

baksmali list dex blah.odex

**Comment by JesusFreke** [comment]
I just pushed my current changes to the v2.2_WIP branch, if you want to take a look.

**Comment by bursoft-git** [comment]
OK, I will test it, thanks!

**Comment by bursoft-git** [comment]
JF, when i try to deodex file with baksmali-2.2-dev.jar, i I get an error
```
c:\BatchApkTool\_framework>java -jar baksmali-2.2-dev.jar x -a 17 --inline-table inline.txt am.odex
Exception in thread "main" java.lang.RuntimeException: Invalid method descriptor: inline.txt
        at org.jf.dexlib2.analysis.CustomInlineMethodResolver.parseAndResolveInlineMethod(CustomInlineMethodResolver.java:107)
        at org.jf.dexlib2.analysis.CustomInlineMethodResolver.<init>(CustomInlineMethodResolver.java:80)
        at org.jf.baksmali.DeodexCommand.getOptions(DeodexCommand.java:77)
        at org.jf.baksmali.DisassembleCommand.run(DisassembleCommand.java:224)
        at org.jf.baksmali.Main.main(Main.java:81)
[inline.txt](https://github.com/JesusFreke/smali/files/265744/inline.txt)

```
this inline.txt works fine with baksmali-2.1.2

**Comment by bursoft-git** [comment]
JesusFreke, how can I deodex /system/framework/framework.jar:classes2.dex from boot-framework.oat?

list dex files from boot-framework.oat
```
java -jar baksmali-2.2-dev.jar l d boot-framework.oat
/system/framework/framework.jar
/system/framework/framework.jar:classes2.dex
```
deodex /system/framework/framework.jar:classes2.dex
```
java -jar baksmali-2.2-dev.jar x boot-framework.oat:/system/framework/framework.jar:classes2.dex
Can't find the file boot-framework.oat:/system/framework/framework.jar:classes2.dex
```

**Comment by JesusFreke** [comment]
Both issues should be fixed in the v2.2_WIP branch now

**Comment by bursoft-git** [comment]
Yes, commit https://github.com/JesusFreke/smali/commit/bd349c1267ebbef2222a2985fa2ead0a95603390 fixed both issues, thanks!

Another question. I deodex boot-apache-xml.oat from Android N (http://developer.android.com/intl/ru/preview/download.html), and got an error:

```
c:\BatchApkTool>java -jar baksmali-2.2-dev.jar x -d "_framework\arm" "_framework\arm\boot-apache-xml.oat:apache-xml.jar"

Error occurred while disassembling class Lorg.apache.xml.utils.Context2; - skipping class
org.jf.util.ExceptionWithContext: Invalid type: V
	at org.jf.dexlib2.analysis.RegisterType.getRegisterType(RegisterType.java:238)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeMoveResult(MethodAnalyzer.java:1087)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyzeInstruction(MethodAnalyzer.java:604)
	at org.jf.dexlib2.analysis.MethodAnalyzer.analyze(MethodAnalyzer.java:200)
	at org.jf.dexlib2.analysis.MethodAnalyzer.<init>(MethodAnalyzer.java:130)
	at org.jf.baksmali.Adaptors.MethodDefinition.addAnalyzedInstructionMethodItems(MethodDefinition.java:463)
	at org.jf.baksmali.Adaptors.MethodDefinition.getMethodItems(MethodDefinition.java:371)
	at org.jf.baksmali.Adaptors.MethodDefinition.writeTo(MethodDefinition.java:238)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeDirectMethods(ClassDefinition.java:281)
	at org.jf.baksmali.Adaptors.ClassDefinition.writeTo(ClassDefinition.java:111)
	at org.jf.baksmali.Baksmali.disassembleClass(Baksmali.java:136)
	at org.jf.baksmali.Baksmali.access$000(Baksmali.java:43)
	at org.jf.baksmali.Baksmali$1.call(Baksmali.java:60)
	at org.jf.baksmali.Baksmali$1.call(Baksmali.java:58)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```
and %errorlevel% = **-1**

What does it mean?

**Comment by JesusFreke** [comment]
Yeah, N deodexing isn't fully working yet

**Comment by bursoft-git** [comment]
I believe that you fix N deodexing )
I do not fully understand
Usually, the errorlevel returns a positive value
I use "if errorlevel >=1" check for errors, should i rewrite my scripts for negative errorlevels?

**Comment by JesusFreke** [comment]
This is fixed in 2.2, which has the "list dex" command
